lib/posix-process/wait.c

```diff
@@ -1,78 +1,250 @@
+		/* Find process with changed state. No change means that
+		 * the state change happened to a different child, so we
+		 * loop back to the semaphore.
+		 */
+		if (pid > 0)
+			waitable = pprocess_state_changed(child, opt);
+		else /* any */
+			waitable = pprocess_child_state_changed(opt);
+
+		if (waitable) {
+			ret = waitable->pid;
+			if (!(opt & WNOWAIT))
+				pprocess_reap(waitable, si);
+			goto out;
+		}
+
+		if (opt & WNOHANG) {
+			ret = 0;
+			break;
+		}
+
+		/* Wait for a state change on one of our children */
+		uk_semaphore_down(&this_thread->process->wait_semaphore);
+	}
+
+out:
+	this_thread->state = POSIX_THREAD_RUNNING;
+	return ret;
+}
 
+UK_SYSCALL_R_DEFINE(pid_t, wait4, pid_t, pid,
+		    int *, wstatus, int, options,
+		    struct rusage * __unused, rusage)
+{
+	siginfo_t si;
+	int ret;
+
+	if (options)
+		if (unlikely(!(options & (WNOHANG | WUNTRACED | WCONTINUED))))
+			return -EINVAL;
+
+	if (uk_list_empty(&uk_pprocess_current()->children))
+		return -ECHILD;
+
+	if (pid < -1 || pid == 0) /* process groups */
+		return -ENOTSUP;
+
+	ret = pprocess_wait(pid, &si, options | WEXITED);
+
+	if (ret > 0 && wstatus) {
+		if (si.si_status == CLD_KILLED)
+			*wstatus = si.si_code;
+		else /* CLD_EXITED */
+			*wstatus = si.si_code << 8;
+	}
+
+	return ret > 0 ? si.si_pid : ret;
+}
+
+UK_LLSYSCALL_R_DEFINE(int, waitid, idtype_t, idtype, id_t, id,
+		      siginfo_t *, infop, int, options,
+		      struct rusage * __unused, rusage)
+{
+	int ret;
+
+	if (options)
+		if (unlikely(!(options & (WEXITED | WSTOPPED | WCONTINUED |
+					  WNOHANG | WNOWAIT))))
+			return -EINVAL;
+
+	if (uk_list_empty(&uk_pprocess_current()->children))
+		return -ECHILD;
+
+	switch (idtype) {
+	case P_PID:
+		ret = pprocess_wait(id, infop, options);
+		break;
+	case P_PIDFD:
+	case P_PGID:
+		ret = -ENOTSUP;
+		break;
+	case P_ALL:
+		ret = pprocess_wait(-1, infop, options);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* POSIX.1-2008 requires that infop must not be NULL, yet Linux
+	 * violates the standard and returns the pid of the waited child
+	 * when that happens. wait(2) discourages applications to rely on
+	 * that behavior, still for backwards compatibility we implement
+	 * Linux's behavior nevertheless.
+	 */
+	return (infop && ret >= 0) ? 0 : ret;
+}
+#else /* !CONFIG_LIBPOSIX_PROCESS_MULTIPROCESS */
 UK_SYSCALL_R_DEFINE(pid_t, wait4, pid_t, pid,
 		    int *, wstatus, int, options,
 		    struct rusage *, rusage)
 {
 	return -ECHILD;
 }
 
-UK_SYSCALL_R_DEFINE(int, waitid, idtype_t, idtype, id_t, id,
-		    siginfo_t *, infop, int, options)
+UK_LLSYSCALL_R_DEFINE(int, waitid, idtype_t, idtype, id_t, id,
+		      siginfo_t *, infop, int, options,
+		      struct rusage *, rusage)
 {
 	return -ECHILD;
 }
+#endif /* !CONFIG_LIBPOSIX_PROCESS_MULTIPROCESS */
 
 #if UK_LIBC_SYSCALLS
 pid_t wait3(int *wstatus, int options, struct rusage *rusage)
 {
-	return uk_syscall_e_wait4((long) -1, (long) wstatus,
-				  (long) options, (long) rusage);
```

Comment:
The `-1`'s from here maybe should be `UK_PID_WAIT_ANY`?

Reply:
See https://github.com/unikraft/unikraft/pull/1630#discussion_r2035688158